byron/ledger/executable-spec/test/Ledger/Delegation/Properties.hs

```diff
@@ -259,18 +271,39 @@ instance HasTrace DBLOCK where
         n <- Gen.integral (Range.linear 0 13)
         pure $! Set.fromAscList $ mkVKeyGenesis <$> [0 .. n]
 
-  sigGen _ (env, st) =
-    DBlock <$> nextSlotGen <*> sigGen @DELEG env st
+  sigGen (Just InvalidDelegationCertificate) _ (env, _st) =
+    DBlock <$> nextSlotGen env <*> Gen.list (Range.constant 0 10) randomDCertGen
     where
-      -- We want the resulting trace to include a large number of epoch
-      -- changes, so we generate an epoch change with higher frequency.
-      nextSlotGen =
-        incSlot <$> Gen.frequency
-                      [ (1, Gen.integral (Range.constant 1 10))
-                      , (2, pure $! slotsPerEpoch (_dSEnvK env))
-                      ]
+      -- | Generate a random delegation certificate, which has a high probability of failing since
+      -- we do not consider the current delegation state. So for instance, we could generate a
+      -- delegation certificate for a genesis key that already delegated in this epoch.
+      randomDCertGen :: Gen DCert
+      randomDCertGen = do
+        (vkg, vk, e) <- (,,) <$> vkgGen' <*> vkGen' <*> epochGen'
+        pure $! mkDCert vkg (signWithGenesisKey vkg (vk, e)) vk e
         where
-          incSlot c = (env ^.slot) `addSlot` SlotCount c
+          vkgGen' = Gen.element $ Set.toList allowed
+          allowed = _dSEnvAllowedDelegators env
+          vkGen' = Gen.element $ VKey . Owner <$> [0 .. (2 * fromIntegral (length allowed))]
+          epochGen' =  Epoch
+                    .  fromIntegral -- We don't care about underflow. We want to generate large epochs anyway.
+                    .  (fromIntegral n +)
+                   <$> Gen.integral (Range.constant (-2 :: Int) 2)
+            where Epoch n = _dSEnvEpoch env
+  sigGen _ _ (env, st) =
```

Comment:
As an example of how we might work the goblins in here, we could add another function definition, something like:
```
  sigGen (Just OtherFailure) _ (env, _st) =
    let genome = $(readGenome "goblin_genomes/OtherFailure.txt") :: [Bool]
        gd = spawnGoblin genome TM.empty
        validGen = sigGen Nothing env st
     in evalState (tinker validGen) gd
```
(excuse my rusty TH)

So we would recur to the valid generator, then apply the goblin to that.

Reply:
Oh, great! I'm looking forward to seeing goblins hooked up to our generators :)